[HYEON] STEP 10 테스트 코드 작성

antaehyeon opened the pull request:
1. jest 공식문서를 보고 개발하였습니다.
2. transfile 중 Babel 을 이용하여 개발하였습니다.
3. MVP 에 대한 테스트를 다양하게 적용하려고 시도하였습니다.
4. STEP10 에 대한 개발로그를 작성하였습니다.

antaehyeon committed:
Feat: MV* 적용 및 동전 투입 버튼 기능 구현

- MVC 구조를 적용하였습니다.
- View에서 동전투입 버튼에 Click Event Listener 를 등록하는 코드를 구현하였습니다.
- 동전을 클릭하면 log 형태로 찍히도록 구현하였습니다 (STEP4)

antaehyeon committed:
Docs: STEP4 개발log 추가

ES6 Classes (by MDN) 내용도 포함되어 있습니다

antaehyeon committed:
Feat: 돈을 투입하는 기능(메서드 단) 구현

MVC 인자로 제대로 넘기지 않는 문제 수정
Model 에 데이터(돈)를 저장하고 불러오도록 구현
VIEW 쪽에서 페이지에 돈을 보여지게 하는 부분 구현

antaehyeon committed:
Refactor: 지갑에 돈을 투입하는 메서드 네이밍 수정

antaehyeon committed:
Feat: 자판기에 돈을 투입하는 기능 추가

antaehyeon committed:
Refactor: 중복되는 코드 메서드로 처리

돈을 표시하는 노드에서 textContent를 수정하는 부분이 중복되어 메서드로 따로 처리함

antaehyeon committed:
Feat: 돈을 투입할 때 마이너스 되는 부분을 막는 기능 추가

돈이 부족할 경우 경고창을 띄웠습니다.

antaehyeon committed:
Docs: STEP5 개발로그 추가

antaehyeon committed:
Refactor: JSDoc 적용

antaehyeon committed:
Refactor: 피드백 적용

querySelector 따로 함수로 구성
인덴트 줄이기

antaehyeon committed:
Docs: README 에 STEP5 피드백 내용 추가

antaehyeon committed:
Docs: STEP5 피드백내용 및 VIEW 단락 추가

antaehyeon committed:
Refactor: VIEW 분리

antaehyeon committed:
Feat: 돈을 투입하면 로그창에 찍히도록 구현

antaehyeon committed:
Feat: 로그창 스크롤 속성 구현

antaehyeon committed:
Fix: 돈이 투입되지 않았는데, 로그가 출력되는 점 수정

antaehyeon committed:
Feat: 로그 텍스트 규격 설정

insertAdjacentHTML 메서드 대신 createElement 와 innerText 속성이용
로그 앞에 모드를 구분함 (Ex. [투입] .. 등)

antaehyeon committed:
Feat: 금액 투입시, 선택할 수 있는 상품 하이라이트 기능 구현

antaehyeon committed:
Docs: STEP6 개발로그 추가

antaehyeon committed:
Refactor: 불필요한 작업 최소화

하이라이트 표시할 때, 객체의 innerText 를 반복적으로 찾는 부분을 Model의 데이터를 이용해 최소화 하였습니다.

antaehyeon committed:
Refactor: 불필요한 반복문 최소화

하이라이트 표시할 때, 이미 하이라이트 된 노드를 찾아서 return 하였습니다.

antaehyeon committed:
Style: 코드 불필요한 부분 삭제

antaehyeon committed:
Docs: STEP6 이슈사항 내용 추가

하이라이트 부분에서 불필요한 부분 삭제하는 과정 내용추가
CSS Multiple class 적용 순위 이슈내용 추가

antaehyeon committed:
Merge branch 'antaehyeon' into STEP6

antaehyeon committed:
Merge pull request #1 from antaehyeon/STEP6

merge Project

antaehyeon committed:
Refactor: STEP6 피드백 적용 및 STEP7 초기화

antaehyeon committed:
Feat: 상품 번호 버튼에 클릭이벤트 등록

이벤트가 정상적으로 클릭되는지 확인하기 위해 alert 기능을 이용하였습니다

antaehyeon committed:
Feat: Model에 데이터를 기반으로 상품번호 클릭이벤트 작동

Model 에 선택한 번호를 저장합니다.
Model 에서 데이터가 정상적으로 저장되는 것을 확인했습니다.

antaehyeon committed:
Feat: Timer 기반 구현

상품 번호를 선택할 때, Timer 작동
상품 번호를 선택한 후, 다른 상품 번호를 선택할 때 Timer Clear 및 재작동
로그를 통해 정상적으로 출력되는 것 확인

antaehyeon committed:
Merge remote-tracking branch 'upstream/antaehyeon' into antaehyeon

antaehyeon committed:
Merge branch 'antaehyeon' of https://github.com/antaehyeon/javascript-vm into antaehyeon

# Conflicts:
#	MVC/model.js
#	MVC/view-util.js
#	MVC/view.js

antaehyeon committed:
Merge branch 'antaehyeon' of https://github.com/antaehyeon/javascript-vm into antaehyeon

# Conflicts:
#	MVC/controller.js
#	MVC/view.js
#	util.js

antaehyeon committed:
Merge branch 'antaehyeon' of https://github.com/antaehyeon/javascript-vm into antaehyeon

# Conflicts:
#	main.html
#	util.js

antaehyeon committed:
Init: STEP10 을 위한 레파지토리 초기화 및 README 작성

antaehyeon committed:
Fix: push 한 node_modules 제거를 위한 push 취소

antaehyeon committed:
Feat: Jest 를 이용한 util.js 테스트코드(util.test.js) 추가

antaehyeon committed:
Feat: util.test.js 기능 완성

antaehyeon committed:
Feat: logView 및 logPresenter 에 대한 testCode 추가

add /Test/log.test.js
mock example test code 추가

antaehyeon committed:
Feat: controlView 에 대한 테스트코드 작성(control.test.js)

antaehyeon committed:
Feat: jest.mock 을 이용한 controlPresenter 테스트 코드 추가

모든 함수가 mock 함수로 변경되어, 제대로 된 테스트가 불가능함

antaehyeon committed:
Feat: control.test.js 테스트코드 완성

antaehyeon committed:
Feat: item.test.js 테스트 코드 추가

antaehyeon committed:
Feat: wallet.test.js 테스트 코드 추가

antaehyeon committed:
Docs: STEP10 개발로그 추가

nigayo commented:
일단 커밋로그는 무척 좋아요.

nigayo commented on Test/util.test.js:
```diff
@@ -0,0 +1,179 @@
+    expect(receivedData).toEqual(expectedResult);
+});
+
+test("상품 선택이 올바른지 테스트합니다", () => {
```
설명은 좀더 구체적일수록 좋아요.
나중에 테스트결과를 쭉 살펴볼때 test함수의 첫번째 문장들만 모아 봐도 무슨테스트인지 알수 있을정도로요.

antaehyeon replied:
구체적으로.. 알겠습니다 👍

nigayo commented on Test/wallet.test.js:
```diff
@@ -0,0 +1,104 @@
+// 해당 클릭이벤트도 어떻게 테스트하는지 의문이 듬
+test("동전을 투입하는 버튼에 이벤트를 등록하는지 테스트합니다", () => {
+    // given
+    walletView.registerClickEventToInsertMoneyBtn = jest.fn();
```
mock을 너무 쉽게 쓰지 않도록 해보세요.
결국 테스트는 진짜를 테스트해야하고, 그럴 수 있는 부분은 그렇게 해보세요.

antaehyeon replied:
확인하였습니다.
그러면 click Event 에서는 보통 어떤식으로 테스트하나요?

crongro reviewed (changes requested):
수고하셨어요.
MVP를 테스트까지 하면서 더 이해가 높아졌기를 바랄게요.
잘했는데요, MOCK테스트를 조금 줄이는 방향으로 개선하면 좋겠어요.

antaehyeon commented on .babelrc:
```diff
@@ -0,0 +1,3 @@
+{
+  "presets": ["es2015"]
```
- Babel is a toolchain that is mainly used to convert ECMAScript 2015+ code into a backwards compatible version of JavaScript in current and older browsers or environments.
- ES6/ES7 코드를 ECMAScript5 코드로 변환하기 위한 도구
- 다양한 작은 모듈들을 담는 일종의 `상자` 역할
- 코드를 컴파일하기 위해 작은 모듈들(presets)을 사용
- `babel-cli` 는 command line 를 사용할 수 있게 지원
- `.babelrc` 는 babel을 설정하기 위한 파일
- `proposal spec` 들이 존재함. 5개의 stage로 구분 (babel-preset-stage-0, 1, 2, 3, 4)
  - babel-preset-stage-4 = babel-preset-es2015
  - ~~한번에 사용할 수 있게 제공하는 것이 `babel-preset-env`~~
  - babel-preset-env 는 babel-preset-2015, 2016, 2017 를 대처한 것
- `babel-polyfill` 은 ES6 환경을 제공
  - polyfill 이란 code 조각으로 런타임에 존재하지 않는 native API의 복사본을 뜻함

antaehyeon commented on Model/VendingMachineModel.js:
```diff
@@ -0,0 +1,165 @@
+     * 상품검증 타이머의 ID를 지정합니다
+     * @param {number} timerFunc - timer ID
+     */
+    setProductVerificationTimerID(id) {
```
VendingMachineControlPresenter:43 에서 사용중이긴 한데, 모든 getter/setter가 필수로 필요하지 않다는 점 기억하겠습니다!

antaehyeon commented on Presenter/VendingMachineWalletPresenter.js:
```diff
@@ -0,0 +1,75 @@
+    this.walletView.registerClickEventToInsertMoneyBtn();
+}
+
+VendingMachineWalletPresenter.prototype = {
```
prototype은 효과적으로 동작한다. new 를 통해 생성된 객체(인스턴스)들이 여러개 존재한다 하더라도, prototype에 연결된 객체들은 동일한 메모리 공간에서 효율적으로 재사용된다. 즉, 두 객체의 prototype 은 같으며, prototype 객체는 최상위 객체까지 연결되어 있어서, prototype 연결고리를 통해 객체간의 상속관계를 만들 수 있다.

- 출처 : 내 블로그 [STEP9](https://antaehyeon.github.io/devlog/2018/08/14/STEP9/)

antaehyeon commented on README/step10.md:
```diff
@@ -0,0 +1,168 @@
+
+   `Jest`
+
+   Jest는 Facebook에서 React 애플리케이션을 포함한 모든 JavaScript 코드를 테스트하는 데 사용되며, Jest의 철학 중 하나는 통합 된 "zero-configuration"경험을 제공하는 것이다. 우리는 엔지니어가 즉시 사용 가능한 도구를 제공받을 때 더 많은 테스트를 작성하게되고 결과적으로 더 안정적이고 건강한 코드 기반을 얻게된다는 것을 알게되었다.
```
이런 좋은 도구를 경험할 수 있게 해주셔서 감사합니다.

antaehyeon commented on Test/control.test.js:
```diff
@@ -0,0 +1,187 @@
+// jest.mock('../Presenter/VendingMachineControlPresenter.js');
+// jest.mock('../View/VendingMachineLogView.js');
+
+const model = new Model();
```
참고하겠습니다.

antaehyeon commented on Test/control.test.js:
```diff
@@ -0,0 +1,187 @@
+// });
+
+// 이런 이벤트들을 어떤 부분을 테스트 해야하는지 잘 감이 안온다.
+test("상품을 선택하는 버튼에 이벤트가 등록되는지 테스트한다", () => {
```
- `describe(name, fn)` creates a block that groups together several related tests in one "test suite".

antaehyeon commented on Test/control.test.js:
```diff
@@ -0,0 +1,187 @@
+});
+
+test("상품을 클릭하는 버튼의 배열을 정상적으로 생성하는지 테스트한다", () => {
+    // given
```
given - when - then 구조에 맞춰서 적용하려고 하니 테스트가 좀 더 수월했던 것 같습니다.
같은 구조에서도 경계가 모호한 테스트들이 있는데, 나중에 질문드리겠습니다!

antaehyeon commented on Test/control.test.js:
```diff
@@ -0,0 +1,187 @@
+    const node = document.createElement('div');
+
+    // when
+    const receivedResult = controlView.changeMoneyNodeTextContent(node, 10000);
```
이런 것들을 magic number 이라고 했던가요..
사용을 최대한 자제하겠습니다 👍

antaehyeon commented on Test/control.test.js:
```diff
@@ -0,0 +1,187 @@
+*/
+test("Model 에서 정상적으로 데이터를 가져와서 설정하는지 테스트합니다", () => {
+    // given
+    model.increaseInvestedMoney(7700);
```
확인하였습니다 👍

antaehyeon commented on Test/control.test.js:
```diff
@@ -0,0 +1,187 @@
+
+    // then
+    expect(setTimeout).toHaveBeenCalledTimes(1);
+    expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 1000);
```
감사합니다 👍

antaehyeon commented on Test/control.test.js:
```diff
@@ -0,0 +1,187 @@
+test("상품을 정상적으로 구매할 수 있는지 테스트합니다", () => {
+    // given
+    const productID = 1;
+    const price = 1000;
```
개인적으로 느꼈을 때, 천단위가 넘어가는 부분들은 제가 magic number 로 사용했던 것 같습니다.
피드백을 통해 앞으로 magic number 의 사용을 자제 하겠습니다

antaehyeon committed:
test: describe 함수 동작 테스트

antaehyeon committed:
Refactor: Magic Number 제거

antaehyeon committed:
Test: Jest, DOM Content, BeforeEach, BeforeAll ...

1. Jest Test를 진행할 때, Document 객체 추가방법
2. Jest Test시, querySelector 동작 테스트
3. Element 출력 테스트
4. beforeEach, beforeAll 동작 테스트

antaehyeon committed:
Refactor: LogView 중 displayLog 메서드가 정확하게 테스트하도록 변경

antaehyeon commented on Test/log.test.js:
```diff
@@ -0,0 +1,121 @@
+    expect(logView.displayLog.name).toEqual("displayLog");
+});
+
+it('로그 노드(DIV)를 정상적으로 삽입하는지 테스트합니다', () => {
```
수정하였습니다 👍

antaehyeon commented on Test/log.test.js:
```diff
@@ -0,0 +1,121 @@
+
+    // then
+    // expect(node.style.visibility).toBe('hidden');
+    expect(logView.setNodeVisibility).toHaveBeenCalled();
```
수정하였습니다 👍

antaehyeon commented on Test/util.test.js:
```diff
@@ -0,0 +1,179 @@
+    // then
+    expect(inputModeResult).toBe("100원이 투입됨");
+    expect(selectModeResult).toBe("100번이 선택됨");
+    expect(refundModeResult).toBe("100원이 반환됨");
```
수정하였습니다 👍

antaehyeon committed:
Docs: Mock 함수로 가짜 테스트하던 코드들 수정

antaehyeon committed:
Refactor: 테스트 제목 자세하게 수정

antaehyeon committed:
Docs: STEP10 개발로그 및 코드정리
